app/controller.go

```diff
@@ -121,12 +121,26 @@ func (app *App) blockBeginner() blockBeginner {
 		gc := getGasCalculator(app.genesisDoc.ConsensusParams)
 		app.Context.deliver = storage.NewState(app.Context.chainstate).WithGas(gc)
 
+		err := ManageVotes(&req, &app.Context, app.logger)
+		if err != nil {
+			app.logger.Error("manage votes error", err)
+		}
+
 		// update the validator set
-		err := app.Context.validators.Setup(req, app.Context.node.ValidatorAddress())
+		err = app.Context.validators.Setup(req, app.Context.node.ValidatorAddress(), app.Context.deliver)
```

Comment:
I have a feeling that by explicitly switching to deliver state, you wanted to see the stake/power update immediately in somewhere else(in query or somewhere). If the penalizing happened in last block, are we able to slash the GUILTY validator's stake/power right in last block ender without delay? Delay can makes things more complicated.